### PR TITLE
Don't warn on array_of_expr with zero length

### PR DESCRIPTION
**Pull request by smowton**
These result from array_set instructions, which when targeting variable-length arrays can end up targeting zero-sized arrays.

Some regression tests in test-gen-support are currently not looking for `warning: ignoring` because of this.

### CODE REVIEW

**Comment by tautschnig** [comment]
Working with zero-sized arrays seems wrong - would it be possible to get a pointer to a regression test?

**Comment by smowton** [comment]
https://github.com/diffblue/cbmc/tree/test-gen-support/regression/cbmc-java/NondetArray2 running against current test-gen-support with latest master merged (e.g. here: https://github.com/smowton/cbmc/tree/smowton/merge/master_20170614)

Basically `new int[0]` is legal Java, and it's an open choice whether we try to intercept that at Java -> array-set or array-set -> array-of or array-of -> prop stage.

**Comment by tautschnig** [comment]
Oh, I was not aware that this was coming from Java. We should certainly have  this backed by a regression test to document the context in which this is expected.

**Comment by smowton** [comment]
Added a test and made the checks a little more precise. I considered avoiding calling `cpp_new[]` at all when the array is zero length, but determined this was unlikely to work as we'd end up with an expression like
`x == 0 ? null : new int[x]`, which I believe would still try to convert the `new int[]` part.

Also I note that https://github.com/diffblue/cbmc/blob/master/src/solvers/flattening/boolbv_if.cpp#L17 already tolerates a zero-length bitvector, so I think it's probably okay to pass these in other places too.

File: src/solvers/flattening/boolbv_array_of.cpp [hunk 1]
```diff
@@ -25,7 +25,14 @@ bvt boolbvt::convert_array_of(const array_of_exprt &expr)
   std::size_t width=boolbv_width(array_type);
 
   if(width==0)
-    return conversion_failed(expr);
+  {
+    // A zero-length array is acceptable;
+    // an element with unknown size is not.
+    if(boolbv_width(array_type.subtype())==0)
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_array_of.cpp]: What about arrays of arrays of zero length? Would Java permit that? If so, this check needs to be amended.

**Comment by tautschnig** [comment]
I'm not sure what's going on here with regard to the build checks...

File: src/solvers/flattening/boolbv_array_of.cpp (hunk 1 above)
Comment (smowton) [inline src/solvers/flattening/boolbv_array_of.cpp]: No, at least in Java such a thing would be an array of pointers

**Comment by smowton** [comment]
Re: checks I accidentally pushed to origin/smowton/... instead of smowton/smowton/... and then deleted the branch created in error. I think the /push and /branch ones relate to that and the PR ones are working as intended.

File: src/solvers/flattening/boolbv_array_of.cpp (hunk 1 above)
Comment (tautschnig) [inline src/solvers/flattening/boolbv_array_of.cpp]: Ok!

**Pull request merged**